pkg/orchestrator/scheduler/types.go

```diff
@@ -104,6 +106,109 @@ func (set execSet) filterFailed() execSet {
 	return set.filterByState(model.ExecutionStateFailed)
 }
 
+// filterOverSubscriptions partitions executions based on if they are more than the desired count.
+func (set execSet) filterByOverSubscriptions(desiredCount int) (execSet, execSet) {
+	remaining := make(execSet)
+	overSubscriptions := make(execSet)
+
+	count := 0
+	for _, exec := range set.ordered() {
+		if count >= desiredCount {
+			overSubscriptions[exec.ComputeReference] = exec
+		} else {
+			remaining[exec.ComputeReference] = exec
+		}
+		count++
+	}
+	return remaining, overSubscriptions
+}
+
+// filterByNodeHealth partitions executions based on their node's health status.
+func (set execSet) filterByNodeHealth(nodeInfos map[string]*model.NodeInfo) (execSet, execSet) {
+	healthy := make(execSet)
+	lost := make(execSet)
+
+	for _, exec := range set {
+		if _, ok := nodeInfos[exec.NodeID]; !ok {
+			lost[exec.ComputeReference] = exec
+			log.Debug().Msgf("Execution %s is running on node %s which is not healthy", exec.ComputeReference, exec.NodeID)
+		} else {
+			healthy[exec.ComputeReference] = exec
+		}
+	}
+	return healthy, lost
+}
```

Comment:
Potential candidates for named returns

Reply:
Nice suggestion. Done in [a274f77](https://github.com/bacalhau-project/bacalhau/pull/2690/commits/a274f777a5a5c220fdce71f78d0936fba186aed1)